C does not handle stack arrays of unknown size

**Describe the bug**
C does not handle stack arrays of unknown size

**To Reproduce**
Provide code to reproduce the behavior:
```python
@pure
@stack_array('tmp')
def test_degree(degree : int):
    from numpy import empty

    tmp = empty(degree+1, dtype=float)
    for i in range(degree+1):
        tmp[i]=0.
    return 1
```

Provide the generated code:
```bash/fortran/c
#include "test.h"
#include <stdint.h>
#include "ndarrays.h"
#include <stdlib.h>

/*........................................*/
int64_t test_degree(int64_t degree)
{
    t_ndarray tmp;
    int64_t i;
    int64_t Out_0001;
    tmp = (t_ndarray){.nd_double=(double[tmp.shape[0]]){},
    .shape=(int64_t[]){tmp.shape[0]},
    .strides=(int64_t[1]){0},
    .nd=1,
    .type=nd_double,
    .is_view=false};
    stack_array_init(&tmp);

    for (i = 0; i < degree + 1; i += 1)
    {
        tmp.nd_double[get_index(tmp, i)] = 0.0;
    }
    Out_0001 = 1;
    return Out_0001;
}
/*........................................*/
```

The compiler says:
```
/home/emily/Code/test/tmp/__pyccel__/test.c: In function ‘test_degree’:
/home/emily/Code/test/tmp/__pyccel__/test.c:12:34: error: compound literal has variable size
     tmp = (t_ndarray){.nd_double=(double[tmp.shape[0]]){},
                                  ^
/home/emily/Code/test/tmp/__pyccel__/test.c:12:56: error: empty scalar initializer
     tmp = (t_ndarray){.nd_double=(double[tmp.shape[0]]){},
                                                        ^
/home/emily/Code/test/tmp/__pyccel__/test.c:12:56: note: (near initialization for ‘(anonymous)’)
```

**Language**
C

--- a/pyccel/codegen/printing/ccode.py
+++ b/pyccel/codegen/printing/ccode.py
@@ -41,6 +41,8 @@ from pyccel.ast.variable import ValuedVariable
 from pyccel.ast.variable import PyccelArraySize, Variable, VariableAddress
 from pyccel.ast.variable import DottedName
 
+from pyccel.ast.sympy_helper import pyccel_to_sympy
+
 
 from pyccel.codegen.printing.codeprinter import CodePrinter
 
@@ -313,15 +315,34 @@ class CCodePrinter(CodePrinter):
         rhs = expr.rhs
         lhs = expr.lhs
         code_init = ''
+        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+
         if lhs.is_stack_array:
-            declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-            length = '*'.join(self._print(i) for i in lhs.alloc_shape)
-            buffer_array = "({declare_dtype}[{length}]){{}}".format(declare_dtype = declare_dtype, length=length)
+            symbol_map = {}
+            used_names_tmp = self._parser.used_names.copy()
+            sympy_shapes = [pyccel_to_sympy(s, symbol_map, used_names_tmp) for s in lhs.alloc_shape]
+
+            length = functools.reduce(operator.mul, sympy_shapes)
+            length_code = '*'.join(self._print(i) for i in lhs.alloc_shape)
+
+            if length.is_constant():
+                buffer_array = "({declare_dtype}[{length}]){{}}".format(
+                                        declare_dtype = declare_dtype,
+                                        length=length_code)
+            else:
+                dummy_array_name, _ = create_incremented_string(self._parser.used_names,
+                                                                prefix = lhs.name+'_data')
+                code_init += "{dtype} {name}[{length}];\n".format(
+                        dtype  = declare_dtype,
+                        name   = dummy_array_name,
+                        length = length_code)
+                buffer_array = dummy_array_name
+
             code_init += self._init_stack_array(expr, buffer_array)
+
         if rhs.fill_value is not None:
             if isinstance(rhs.fill_value, Literal):
-                dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-                code_init += 'array_fill(({0}){1}, {2});\n'.format(dtype, self._print(rhs.fill_value), self._print(lhs))
+                code_init += 'array_fill(({0}){1}, {2});\n'.format(declare_dtype, self._print(rhs.fill_value), self._print(lhs))
             else:
                 code_init += 'array_fill({0}, {1});\n'.format(self._print(rhs.fill_value), self._print(lhs))
         return code_init

--- a/pyccel/codegen/printing/ccode.py
+++ b/pyccel/codegen/printing/ccode.py
@@ -315,7 +315,7 @@ class CCodePrinter(CodePrinter):
         code_init = ''
         if lhs.is_stack_array:
             declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-            length = '*'.join(self._print(i) for i in lhs.shape)
+            length = '*'.join(self._print(i) for i in lhs.alloc_shape)
             buffer_array = "({declare_dtype}[{length}]){{}}".format(declare_dtype = declare_dtype, length=length)
             code_init += self._init_stack_array(expr, buffer_array)
         if rhs.fill_value is not None:
@@ -343,15 +343,15 @@ class CCodePrinter(CodePrinter):
         lhs = expr.lhs
         rhs = expr.rhs
         dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
-        shape = ", ".join(self._print(i) for i in lhs.shape)
+        shape = ", ".join(self._print(i) for i in lhs.alloc_shape)
         declare_dtype = self.find_in_dtype_registry('int', 8)
 
         shape_init = "({declare_dtype}[]){{{shape}}}".format(declare_dtype=declare_dtype, shape=shape)
         strides_init = "({declare_dtype}[{length}]){{0}}".format(declare_dtype=declare_dtype, length=len(lhs.shape))
         if isinstance(buffer_array, Variable):
             buffer_array = "{0}.{1}".format(self._print(buffer_array), dtype)
-        cpy_data = '{0} = (t_ndarray){{.{1}={2},\n .shape={3},\n .strides={4},\n '
-        cpy_data += '.nd={5},\n .type={1},\n .is_view={6}}};\n'
+        cpy_data = '{0} = (t_ndarray){{\n.{1}={2},\n .shape={3},\n .strides={4},\n '
+        cpy_data += '.nd={5},\n .type={1},\n .is_view={6}\n}};\n'
         cpy_data = cpy_data.format(self._print(lhs), dtype, buffer_array,
                     shape_init, strides_init, len(lhs.shape), 'false')
         cpy_data += 'stack_array_init(&{});\n'.format(self._print(lhs))

--- a/tests/pyccel/test_pyccel.py
+++ b/tests/pyccel/test_pyccel.py
@@ -413,9 +413,8 @@ def test_pyccel_calling_directory(language):
     compare_pyth_fort_output( pyth_out, fort_out )
 
 #------------------------------------------------------------------------------
-# C does not handle stack arrays of variable size. See #760
-def test_in_specified():
-    pyccel_test("scripts/runtest_degree_in.py")
+def test_in_specified(language):
+    pyccel_test("scripts/runtest_degree_in.py", language=language)
 
 #------------------------------------------------------------------------------
 @pytest.mark.parametrize( "test_file", ["scripts/hope_benchmarks/hope_fib.py",

